### PR TITLE
Taking a swing at the inline partial shuffle

### PR DESCRIPTION
**Pull request by evs-chris**
### Part 1

~~Per #1838's discussion, this gets inline partials scoped back to their nearest Ractive instance. It does so by simply copying the partials from the element to the element root when the element is instantiated. This does have at least one potential issue that I can think of: non-rendered segments of the template can't contribute inline partials. I don't know if that's a bug or a feature. It kinda makes sense, but then the partial hangs around if the element is unrendered, which seems a little wrong. So, I could go either way on how to handle that:~~

1. ~~Read through the whole template when it is loaded and pull in any partials that don't exist down a component tree. This is roughly the same as the previous behavior.~~
2. ~~When the element is torn down, reset any partials that it contributed at init back to what they were if they haven't been updated in the interim. For partials unique to the element, this would remove them, and for partials that override, the overridden partial would be restored.~~

~~I'm leaning toward option 2 at the moment. It seems to track better with part 2 of this.~~

After discussion, it seems that just leaving partials attached to their elements and looking for partials up the fragment hierarchy is a more consistent approach.

### Part 2

Per @martypdx's suggestion, this also lets `getPartialTemplate` look up the element hierarchy for a partial before checking the usual places. This is pretty neat for doing partial overrides outside of components.

~~Something to consider here is how this should interact with `resetPartial`. Should all instances get touched, or should partials that were originally rendered from inlines end up with their inline again, or something else?~~

The current behavior is, upon reset, partials with inline overrides reset to their nearest inline. That seems to be most consistent with inline partials as full-fledged members of a template. Inline partials at root levels can still be reset, as they are treated as normal partials.

### CODE REVIEW

**Comment by evs-chris** [comment]
Having pondered it a bit more, I favor dropping the instance scoping all together for inline partials. It makes more sense to me to have them be scoped only to elements since they are part of the virtual dom.

In that context, it makes sense that `resetPartial` would not touch inline partials unless they are attached to an instance, which, coincidentally, is the current behavior in this PR.

TL;DR (for the whole PR) - I think inline partials at a root should behave as normal partials; inline partials at an element should only affect fragments further down the hierarchy and should be left otherwise unmolested

**Comment by martypdx** [comment]
If I understand this, element partials get copied to root so they are generally available. However, a partial use will look up the vdom elements and favor partials of closest ancestor. Best of both worlds?

**Comment by evs-chris** [comment]
That's what the PR does as it stands, but there are some potentially weird issues with copying the partials. For instance
```js
new Ractive({
  el: 'main',
  template: '{{#foo}}<div>{{#partial foo}}foo{{/partial}}</div>{{/}}<button on-click="toggle(\'foo\')">Go</button>{{>foo}}'
});
```
'foo' does not appear until you click the button the first time, but it remains after you click it again. So if there was already a 'foo' partial, it would be overwritten and remain so if the element that overwrote it disappeared. This could have particularly strange implications for elements in iterative sections. I suppose another solution to that would be to keep element partials from overwriting existing root partials?

**Comment by martypdx** [comment]
Or we could respect "scope", which is a fairly standard model.

```html
{{#foo}}
  <div>
    {{#partial foo}}foo{{/partial}}
    {{>foo}} <!-- this find foo -->
  </div>
{{/}}
<button on-click="toggle(\'foo\')">Go</button>
{{>foo}}  <!-- causes warning because no foo here -->
```

Then no need to "transfer" to root, either you find the partial, or not, based on scope.

**Comment by martypdx** [comment]
Just to add, I can't think of a case where I nested a partial in a sibling of where it was used. Maybe it's just me, but it would bug me to see it nested outside of where it was used.

**Comment by evs-chris** [comment]
> Or we could respect "scope", which is a fairly standard model.

That's what I was thinking too. I've updated the PR.

**Comment by Rich-Harris** [comment]
101379-jim-carrey-i-like-it-alot-gif-gflm
